### PR TITLE
Add a new layer: Global Average Pooling

### PR DESCRIPTION
**Pull request by karandesai-96**
Global Average Pooling layer was introduced in the paper "Network in Network" (2013). It is an alternative to keeping fully connected layers at the end of a deep convolutional neural network.

Global Average Pooling layer has been used in recent architectures such as Inception Network, ResNet and Xception Network. It performs channel wise pooling and hence a layer with `c` channels will produce an output of `c` units, which is usually followed by softmax, giving classification scores.

#### Figure:
image

#### Content in this PR:

- [x] Add forward and backward op kernels for internal backend.
- [x] Add a class for global average pooling params.
- [x] Layer class implementation ( `class global_average_pooling_layer : public layer` ).
- [x] Relevant tests.
- [x] Serialization helpers and their tests.

#### References:

1. **"Network In Network"** by Min Lin, Qiang Chen, Shuicheng Yan, 2013. ( [arXiv:1312.4400v3 cs.NE](https://arxiv.org/abs/1312.4400v3) ) 
2. **"Going Deeper with Convolutions"** by Szegedy et. al., 2014. ( [arXiv:1409.4842 cs.CV](https://arxiv.org/abs/1409.4842) ).

### CODE REVIEW

File: test/test_global_average_pooling_layer.h [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+  vec_t res      = l.forward({{in}})[0][0];
+
+  EXPECT_FLOAT_EQ(expected[0], res[0]);
+}
+
+TEST(global_ave_pool, setup_internal) {
+  global_average_pooling_layer l(4, 4, 1, core::backend_t::internal);
+
+  EXPECT_EQ(l.parallelize(), true);              // if layer can be parallelized
+  EXPECT_EQ(l.in_channels(), serial_size_t(1));  // num of input tensors
```
Comment (edgarriba) [inline test/test_global_average_pooling_layer.h]: check https://github.com/tiny-dnn/tiny-dnn/pull/667/files#diff-cc023dbe1a164eea581096cc7945bf58R22

File: tiny_dnn/core/params/global_avepool_params.h [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+*/
+#pragma once
+#include "tiny_dnn/core/params/params.h"
+
+namespace tiny_dnn {
+namespace core {
+
+class global_avepool_params : public Params {
+ public:
+  index3d<serial_size_t> in;
```
Comment (edgarriba) [inline tiny_dnn/core/params/global_avepool_params.h]: `size_t`

File: tiny_dnn/layers/global_average_pooling_layer.h [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+  /* Forward and backward ops */
+  std::shared_ptr<core::OpKernel> kernel_fwd_;
+  std::shared_ptr<core::OpKernel> kernel_back_;
+
+  void init_backend(backend_t backend_type) {
+    core::OpKernelConstruction ctx =
+      core::OpKernelConstruction(layer::device(), &params_);
+
+    if (backend_type == backend_t::internal ||
+        backend_type == backend_t::nnpack || backend_type == backend_t::avx) {
```
Comment (edgarriba) [inline tiny_dnn/layers/global_average_pooling_layer.h]: just `backend_t::internal`

File: tiny_dnn/layers/global_average_pooling_layer.h [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+                               backend_t backend_type = core::default_engine())
+    : global_average_pooling_layer(
+        in_shape.width_, in_shape.height_, in_shape.depth_, backend_type) {}
+
+  /**
+   * @param in_width     [in] width of input image
+   * @param in_height    [in] height of input image
+   * @param in_channels  [in] the number of input image channels (depth)
+  **/
+  global_average_pooling_layer(serial_size_t in_width,
```
Comment (edgarriba) [inline tiny_dnn/layers/global_average_pooling_layer.h]: I guess that `serial_size_t` is because of layer members, but we should start to think in fixing this

File: tiny_dnn/layers/global_average_pooling_layer.h [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+  std::string layer_type() const override {
+    return std::string("global-ave-pool");
+  }
+
+  std::pair<serial_size_t, serial_size_t> pool_size() const {
+    return std::make_pair(params_.in.width_, params_.in.height_);
+  }
+
+#ifndef CNN_NO_SERIALIZATION
```
Comment (edgarriba) [inline tiny_dnn/layers/global_average_pooling_layer.h]: not needed anymore https://github.com/tiny-dnn/tiny-dnn/pull/667/files#diff-123f62ed8f57f99ec37c4e4dffe64146L40

File: tiny_dnn/layers/global_average_pooling_layer.h [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+  global_average_pooling_layer(serial_size_t in_width,
+                               serial_size_t in_height,
+                               serial_size_t in_channels,
+                               backend_t backend_type = core::default_engine())
+    : layer({vector_type::data}, {vector_type::data}) {
+    set_global_avepool_params(shape3d(in_width, in_height, in_channels),
+                              shape3d(in_channels, 1, 1));
+
+    init_backend(backend_type);
+    layer::set_backend_type(backend_type);
```
Comment (edgarriba) [inline tiny_dnn/layers/global_average_pooling_layer.h]: it makes more sense to set the backend type inside the `init_backend`routine. Others layers will also need this -> next PR

**Comment by edgarriba** [comment]
@karandesai-96 last comment, it's fine for now to implement the layer with Ops and backends strategy, but for future works, just implement basic forward and backward routines. Refactoring will come later. Just make sure before adding too many lines of code that we can have different backends.
@beru AVX for this?

File: tiny_dnn/layers/global_average_pooling_layer.h [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+ * applies channel-wise global average pooling to spatial data.
+ **/
+class global_average_pooling_layer : public layer {
+ public:
+  using layer::parallelize_;
+
+  global_average_pooling_layer(const shape3d &in_shape,
+                               backend_t backend_type = core::default_engine())
+    : global_average_pooling_layer(
+        in_shape.width_, in_shape.height_, in_shape.depth_, backend_type) {}
```
Comment (Randl) [inline tiny_dnn/layers/global_average_pooling_layer.h]: Do we need separate `width` and `height`? We can generalize to any number of dimensions by passing just area and depth, without caring on number of dimensions. Or passing just depth and deducing area. Or even deducing all the parameters similarly to activations.

**Comment by beru** [comment]
@edgarriba 
> @beru AVX for this?

Writing `vectorize::accumulate_2d` might be good.
It can also be used in internal backend.

File: tiny_dnn/layers/global_average_pooling_layer.h (hunk 3 above)
Comment (karandesai-96) [inline tiny_dnn/layers/global_average_pooling_layer.h]: we need to fall back to internal backend else it will fail a few jobs on travis. similar thing is done for certain other layers.

File: tiny_dnn/layers/global_average_pooling_layer.h [hunk 8]
```diff
@@ -109,7 +109,8 @@ class global_average_pooling_layer : public layer {
       core::OpKernelConstruction(layer::device(), &params_);
 
     layer::set_backend_type(backend_type);
-    if (backend_type == backend_t::internal || backend_type == backend_t::avx || backend_type == backend_t::nnpack) {
+    if (backend_type == backend_t::internal || backend_type == backend_t::avx ||
+        backend_type == backend_t::nnpack) {
```
Comment (edgarriba) [inline tiny_dnn/layers/global_average_pooling_layer.h]: really, I think it can lead to future bugs. To prevent this you can use macros at testing time
https://github.com/tiny-dnn/tiny-dnn/blob/master/test/test_convolutional_layer.h#L212

File: tiny_dnn/layers/global_average_pooling_layer.h (hunk 4 above)
Comment (beru) [inline tiny_dnn/layers/global_average_pooling_layer.h]: #553 is related.

File: tiny_dnn/core/kernels/global_avepool_op.h [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+    in the LICENSE file.
+*/
+#pragma once
+
+#include "tiny_dnn/core/framework/op_kernel.h"
+#include "tiny_dnn/core/kernels/global_avepool_op_internal.h"
+
+namespace tiny_dnn {
+
+class GlobalAvePoolOp : public core::OpKernel {
```
Comment (beru) [inline tiny_dnn/core/kernels/global_avepool_op.h]: This isn't particularly about Global Averaging Pool but,
@edgarriba Why `core::OpKernel` and its descendants are `CamelCase`?
Is it a bad thing or too late to change their names to `snake_case`?
I'm not really a fellow member of coding conventions police though.

File: tiny_dnn/layers/global_average_pooling_layer.h (hunk 7 above)
Comment (beru) [inline tiny_dnn/layers/global_average_pooling_layer.h]: It makes me wonder why we have to always supply dimentional parameters when we construct layers... (I don't take internal implemetation into consideration right now)
#641 is related

The way I see it, this layer doesn't learn weights or sth during back propagation and works as a filter.

File: tiny_dnn/core/kernels/global_avepool_op_internal.h [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+  const core::global_avepool_params &params,
+  const bool layer_parallelize) {
+  for_i(layer_parallelize, in_data.size(), [&](int sample) {
+    const vec_t &in = in_data[sample];
+    vec_t &out      = out_data[sample];
+
+    const size_t pool_area = params.in.width_ * params.in.height_;
+    for (size_t i = 0; i < params.in.depth_; i++) {
+      for (size_t j = 0; j < pool_area; j++) {
+        out[i] += in[i * pool_area + j];
```
Comment (beru) [inline tiny_dnn/core/kernels/global_avepool_op_internal.h]: You should use a local variable when accumulating values.

File: tiny_dnn/core/kernels/global_avepool_op_internal.h [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+      }
+      out[i] /= pool_area;
```
Comment (beru) [inline tiny_dnn/core/kernels/global_avepool_op_internal.h]: Division is very costly so doing that inside a loop wastes a fractional amount of precious electricity .
In many cases, you can replace divisions with reciprocal multiplications by sacrificing code readability.

Comment (karandesai-96) [inline tiny_dnn/core/kernels/global_avepool_op_internal.h]: We had this readability vs micro optimization discussion before, but yes - this is a tight for loop. I can sacrifice readability.

**Comment by bhack** [comment]
https://github.com/ridiculousfish/libdivide/blob/master/README.md

**Comment by beru** [comment]
@bhack `livdivide` is for integer division. So I think it can't be used for floating-point values in place.
And `float / int` operation, divisor `int` type value is automatically converted to `float` type by C++ compiler (but if divisor is literal and it is equal to 2 to the power n, maybe optimized but that's not the case here).

I can be wrong though.

Aside from that, I've noticed that you tend to write only URL so I have to make a wild guess.
The size of my brain is like :chicken: so it'd be great if you could attach your comment when introducing URL.

File: tiny_dnn/core/kernels/global_avepool_op_internal.h (hunk 11 above)
Comment (beru) [inline tiny_dnn/core/kernels/global_avepool_op_internal.h]: There's such a dilemma. And probablly Global Averaging Pooling layer operation may not be the hot spot of NN computation. So it's possible that the performance gain from optimizing this layer isn't very rewardful.

So if we need to organize optimization work, writing profiling feature and benchmark tests (layerwise? opwise?) will probablly aid the work.

**Comment by bhack** [comment]
Yes sorry I've pasted the wrong url from the bookmark. The correct one is this and doesn't require a comment. http://stackoverflow.com/questions/35506226/will-the-compiler-optimize-division-into-multiplication

**Comment by beru** [comment]
@bhack Thank you for the correction.

I guess doing micro-optimization for CPU is like training how to walk faster. 
So in practical, we should rely on hardware accelerators.

File: tiny_dnn/core/kernels/global_avepool_op_internal.h (hunk 11 above)
Comment (Randl) [inline tiny_dnn/core/kernels/global_avepool_op_internal.h]: Have you checked generated assembly? I m 99% sure generated code has this optimisation

Comment (karandesai-96) [inline tiny_dnn/core/kernels/global_avepool_op_internal.h]: Also, we need to check whether these things we care about so much, are already handled by our `-O31 optimization. I am not sure about specific details. http://gcc.gnu.org/onlinedocs/gcc/Optimize-Options.html

Comment (beru) [inline tiny_dnn/core/kernels/global_avepool_op_internal.h]: @Randl I've just confirmed `vdivss` instruction is used with `-mavx -O3` compile options.

Please read the page.
http://stackoverflow.com/a/35506382
https://github.com/tiny-dnn/tiny-dnn/pull/564#discussion_r100667964

File: tiny_dnn/core/kernels/global_avepool_op.h (hunk 9 above)
Comment (edgarriba) [inline tiny_dnn/core/kernels/global_avepool_op.h]: at some point I had some class naming conflicts that I could solve by using `CamelCase` then we came with this convention stuff

File: tiny_dnn/core/kernels/global_avepool_op_internal.h (hunk 11 above)
Comment (Randl) [inline tiny_dnn/core/kernels/global_avepool_op_internal.h]: I see. Then it might make sense to optimize it into single division. @karandesai-96 Do you care to run a simple benchmark?

**Comment by karandesai-96** [comment]
Right so we're relying on hardware accelerators alone. And about @edgarriba's comment - I am adding AVX implementation in next PR.
